server/service/hosts.go

```diff
@@ -332,14 +332,25 @@ func (svc *Service) DeleteHosts(ctx context.Context, ids []uint, filter *map[str
 		}
 
 		mdmLifecycle := mdmlifecycle.New(svc.ds, svc.logger)
+		lifecycleErrs := []error{}
+		serialsWithErrs := []string{}
 		for _, host := range hosts {
 			if fleet.MDMSupported(host.Platform) {
-				err := mdmLifecycle.Do(ctx, mdmlifecycle.HostOptions{
+				if err := mdmLifecycle.Do(ctx, mdmlifecycle.HostOptions{
 					Action:   mdmlifecycle.HostActionDelete,
 					Host:     host,
 					Platform: host.Platform,
-				})
-				return err
+				}); err != nil {
+					lifecycleErrs = append(lifecycleErrs, err)
+					serialsWithErrs = append(serialsWithErrs, host.HardwareSerial)
+				}
+			}
+		}
+		if len(lifecycleErrs) > 0 {
+			msg := fmt.Sprintf("failed to restore pending host records for one or more MDM devices: %+v", serialsWithErrs)
+			return &fleet.BadRequestError{
```

Comment:
Is there a more appropriate error type to return here? Maybe we want to signal 500 error instead.

Reply:
Yeah I think it would be a 500 (likely DB failure), unlikely to be a client error (400 basically means fix your request and try again, I don't think there's much they can change for a delete hosts call?).

Reply:
@mna, is there a particular error struct that you would recommend using in this case? Or just return `ctxerr.Wrap(ctx, errors.Join(lifecycleErrs...), msg)` and let the defaults take care of the rest?

Reply:
I think `errors.Join` will end up in a bare `error` interface that will end up generating a 500 (worth testing to make sure that's correct, but I'm pretty sure that's how it would end up).

Reply:
Thanks, tested locally and confirmed :)